# Data Loading

Import library-library yang akan digunakan

In [47]:
# Import library
import pandas as pd
import numpy as np
import tensorflow as tf

Load dataset dari file csv dan tampilkan dengan pandas dataframe

In [48]:
# Membaca dataset
 
df = pd.read_csv("/content/imdb_top_1000.csv")
df

Poster_Link  \
0    https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1    https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2    https://m.media-amazon.com/images/M/MV5BMTMxNT...   
3    https://m.media-amazon.com/images/M/MV5BMWMwMG...   
4    https://m.media-amazon.com/images/M/MV5BMWU4N2...   
..                                                 ...   
995  https://m.media-amazon.com/images/M/MV5BNGEwMT...   
996  https://m.media-amazon.com/images/M/MV5BODk3Yj...   
997  https://m.media-amazon.com/images/M/MV5BM2U3Yz...   
998  https://m.media-amazon.com/images/M/MV5BZTBmMj...   
999  https://m.media-amazon.com/images/M/MV5BMTY5OD...   

                 Series_Title Released_Year Certificate  Runtime  \
0    The Shawshank Redemption          1994           A  142 min   
1               The Godfather          1972           A  175 min   
2             The Dark Knight          2008          UA  152 min   
3      The Godfather: Part II          1974           A  202 min   
4                12 Angry Men          1957           U   96 min   
..                        ...           ...         ...      ...   
995    Breakfast at Tiffany's          1961           A  115 min   
996                     Giant          1956           G  201 min   
997     From Here to Eternity          1953      Passed  118 min   
998                  Lifeboat          1944         NaN   97 min   
999              The 39 Steps          1935         NaN   86 min   

                        Genre  IMDB_Rating  \
0                       Drama          9.3   
1                Crime, Drama          9.2   
2        Action, Crime, Drama          9.0   
3                Crime, Drama          9.0   
4                Crime, Drama          9.0   
..                        ...          ...   
995    Comedy, Drama, Romance          7.6   
996            Drama, Western          7.6   
997       Drama, Romance, War          7.6   
998                Drama, War          7.6   
999  Crime, Mystery, Thriller          7.6   

                                              Overview  Meta_score  \
0    Two imprisoned men bond over a number of years...        80.0   
1    An organized crime dynasty's aging patriarch t...       100.0   
2    When the menace known as the Joker wreaks havo...        84.0   
3    The early life and career of Vito Corleone in ...        90.0   
4    A jury holdout attempts to prevent a miscarria...        96.0   
..                                                 ...         ...   
995  A young New York socialite becomes interested ...        76.0   
996  Sprawling epic covering the life of a Texas ca...        84.0   
997  In Hawaii in 1941, a private is cruelly punish...        85.0   
998  Several survivors of a torpedoed merchant ship...        78.0   
999  A man in London tries to help a counter-espion...        93.0   

                 Director              Star1              Star2  \
0          Frank Darabont        Tim Robbins     Morgan Freeman   
1    Francis Ford Coppola      Marlon Brando          Al Pacino   
2       Christopher Nolan     Christian Bale       Heath Ledger   
3    Francis Ford Coppola          Al Pacino     Robert De Niro   
4            Sidney Lumet        Henry Fonda        Lee J. Cobb   
..                    ...                ...                ...   
995         Blake Edwards     Audrey Hepburn     George Peppard   
996        George Stevens   Elizabeth Taylor        Rock Hudson   
997        Fred Zinnemann     Burt Lancaster   Montgomery Clift   
998      Alfred Hitchcock  Tallulah Bankhead        John Hodiak   
999      Alfred Hitchcock       Robert Donat  Madeleine Carroll   

              Star3           Star4  No_of_Votes        Gross  
0        Bob Gunton  William Sadler      2343110   28,341,469  
1        James Caan    Diane Keaton      1620367  134,966,411  
2     Aaron Eckhart   Michael Caine      2303232  534,858,444  
3     Robert Duvall    Diane Keaton      1129952   57,300,000  
4     Mart

Gunakan method .info() untuk melihat fitur-fitur pada dataset dan tipe datanya

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Poster_Link    1000 non-null   object 
 1   Series_Title   1000 non-null   object 
 2   Released_Year  1000 non-null   object 
 3   Certificate    899 non-null    object 
 4   Runtime        1000 non-null   object 
 5   Genre          1000 non-null   object 
 6   IMDB_Rating    1000 non-null   float64
 7   Overview       1000 non-null   object 
 8   Meta_score     843 non-null    float64
 9   Director       1000 non-null   object 
 10  Star1          1000 non-null   object 
 11  Star2          1000 non-null   object 
 12  Star3          1000 non-null   object 
 13  Star4          1000 non-null   object 
 14  No_of_Votes    1000 non-null   int64  
 15  Gross          831 non-null    object 
dtypes: float64(2), int64(1), object(13)
memory usage: 125.1+ KB


# Data Understanding dan Data Preparation

Karena sistem rekomendasi ini akan merekomendasikan film berdasarkan genrenya, maka fitur title harus unique dan berkorespondesi satu-satu terhadap fitur Genre. Cek berapa record judul film yang unique

In [50]:
len(df.Series_Title.unique())

999

Dari 1000 record dataset, ternyata ada 999 record dengan judul yang unik. Cek record mana saja yang judulnya tidak unik.

In [51]:
df.loc[df['Series_Title'].duplicated() == True]

Poster_Link Series_Title  \
136  https://m.media-amazon.com/images/M/MV5BYmJhZm...     Drishyam   

    Released_Year Certificate  Runtime                  Genre  IMDB_Rating  \
136          2015          UA  163 min  Crime, Drama, Mystery          8.2   

                                              Overview  Meta_score  \
136  Desperate measures are taken by a man who trie...         NaN   

            Director       Star1         Star2 Star3         Star4  \
136  Nishikant Kamat  Ajay Devgn  Shriya Saran  Tabu  Rajat Kapoor   

     No_of_Votes    Gross  
136        70367  739,478

Terlihat ada 1 judul film yang redundant.

In [52]:
df.loc[df['Series_Title'] == "Drishyam"]

Poster_Link Series_Title  \
87   https://m.media-amazon.com/images/M/MV5BYmY3Mz...     Drishyam   
136  https://m.media-amazon.com/images/M/MV5BYmJhZm...     Drishyam   

    Released_Year Certificate  Runtime                   Genre  IMDB_Rating  \
87           2013           U  160 min  Crime, Drama, Thriller          8.3   
136          2015          UA  163 min   Crime, Drama, Mystery          8.2   

                                              Overview  Meta_score  \
87   A man goes to extreme lengths to save his fami...         NaN   
136  Desperate measures are taken by a man who trie...         NaN   

            Director       Star1         Star2         Star3         Star4  \
87     Jeethu Joseph    Mohanlal         Meena  Asha Sharath        Ansiba   
136  Nishikant Kamat  Ajay Devgn  Shriya Saran          Tabu  Rajat Kapoor   

     No_of_Votes    Gross  
87         30722      NaN  
136        70367  739,478

Dari fakta di atas, dapat kita simpulkan bahwa film yang redundant ternyata adalah film yang memiliki remake version. Oleh karena itu, judul film tsb harus dipadupadankan dengan tahun rilisnya agar bisa menjadi unik. Kemudian, cek kembali jumlah value unik dari fitur Series_Title

In [53]:
for i in range(len(df.index)):
  if (df['Series_Title'] == "Drishyam")[i] == True:
    df["Series_Title"][i] = df['Series_Title'][i] + ' ' + df['Released_Year'][i]

len(df.Series_Title.unique())

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


1000

Sebuah film dapat memiliki lebih dari 1 genre. Karena judul film dan genre harus berkorespondensi satu-satu, maka ekstrak fitur baru (First_genre) yang mewakili genre pertama yang tertera pada fitur Genre.

In [54]:
first_genre = []

for i in range(len(df.index)):
  genre_list = df["Genre"][i].split(", ")
  first_genre.append(genre_list[0])

df["First_genre"] = pd.DataFrame(first_genre)
df

Poster_Link  \
0    https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1    https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2    https://m.media-amazon.com/images/M/MV5BMTMxNT...   
3    https://m.media-amazon.com/images/M/MV5BMWMwMG...   
4    https://m.media-amazon.com/images/M/MV5BMWU4N2...   
..                                                 ...   
995  https://m.media-amazon.com/images/M/MV5BNGEwMT...   
996  https://m.media-amazon.com/images/M/MV5BODk3Yj...   
997  https://m.media-amazon.com/images/M/MV5BM2U3Yz...   
998  https://m.media-amazon.com/images/M/MV5BZTBmMj...   
999  https://m.media-amazon.com/images/M/MV5BMTY5OD...   

                 Series_Title Released_Year Certificate  Runtime  \
0    The Shawshank Redemption          1994           A  142 min   
1               The Godfather          1972           A  175 min   
2             The Dark Knight          2008          UA  152 min   
3      The Godfather: Part II          1974           A  202 min   
4                12 Angry Men          1957           U   96 min   
..                        ...           ...         ...      ...   
995    Breakfast at Tiffany's          1961           A  115 min   
996                     Giant          1956           G  201 min   
997     From Here to Eternity          1953      Passed  118 min   
998                  Lifeboat          1944         NaN   97 min   
999              The 39 Steps          1935         NaN   86 min   

                        Genre  IMDB_Rating  \
0                       Drama          9.3   
1                Crime, Drama          9.2   
2        Action, Crime, Drama          9.0   
3                Crime, Drama          9.0   
4                Crime, Drama          9.0   
..                        ...          ...   
995    Comedy, Drama, Romance          7.6   
996            Drama, Western          7.6   
997       Drama, Romance, War          7.6   
998                Drama, War          7.6   
999  Crime, Mystery, Thriller          7.6   

                                              Overview  Meta_score  \
0    Two imprisoned men bond over a number of years...        80.0   
1    An organized crime dynasty's aging patriarch t...       100.0   
2    When the menace known as the Joker wreaks havo...        84.0   
3    The early life and career of Vito Corleone in ...        90.0   
4    A jury holdout attempts to prevent a miscarria...        96.0   
..                                                 ...         ...   
995  A young New York socialite becomes interested ...        76.0   
996  Sprawling epic covering the life of a Texas ca...        84.0   
997  In Hawaii in 1941, a private is cruelly punish...        85.0   
998  Several survivors of a torpedoed merchant ship...        78.0   
999  A man in London tries to help a counter-espion...        93.0   

                 Director              Star1              Star2  \
0          Frank Darabont        Tim Robbins     Morgan Freeman   
1    Francis Ford Coppola      Marlon Brando          Al Pacino   
2       Christopher Nolan     Christian Bale       Heath Ledger   
3    Francis Ford Coppola          Al Pacino     Robert De Niro   
4            Sidney Lumet        Henry Fonda        Lee J. Cobb   
..                    ...                ...                ...   
995         Blake Edwards     Audrey Hepburn     George Peppard   
996        George Stevens   Elizabeth Taylor        Rock Hudson   
997        Fred Zinnemann     Burt Lancaster   Montgomery Clift   
998      Alfred Hitchcock  Tallulah Bankhead        John Hodiak   
999      Alfred Hitchcock       Robert Donat  Madeleine Carroll   

              Star3           Star4  No_of_Votes        Gross First_genre  
0        Bob Gunton  William Sadler      2343110   28,341,469       Drama  
1        James Caan    Diane Keaton      1620367  134,966,411       Crime  
2     Aaron Eckhart   Michael Caine      2303232  534,858,444      Action  
3     Robert Duvall    Dia

Hapus semua fitur, kecuali Series_Title dan First_genre, karena keduanya akan menjadi fitur pembentuk sistem rekomendasi.

In [55]:
columns = df.columns.tolist()
wanted_columns = ["Series_Title", "First_genre"]

for column in df.columns.tolist():
  if column in wanted_columns:
    columns.remove(column)

df = df.drop(columns=columns)
df

Series_Title First_genre
0    The Shawshank Redemption       Drama
1               The Godfather       Crime
2             The Dark Knight      Action
3      The Godfather: Part II       Crime
4                12 Angry Men       Crime
..                        ...         ...
995    Breakfast at Tiffany's      Comedy
996                     Giant       Drama
997     From Here to Eternity       Drama
998                  Lifeboat       Drama
999              The 39 Steps       Crime

[1000 rows x 2 columns]

Cek missing value dan hapus jika ada

In [56]:
df.isnull().sum()

Series_Title    0
First_genre     0
dtype: int64

# Model Development

Lakukan proses transformasi fitur menggunakan TfidfVectorizer.

In [57]:
from sklearn.feature_extraction.text import TfidfVectorizer
 
# Inisialisasi TfidfVectorizer
tf = TfidfVectorizer()
 
# Melakukan perhitungan idf pada fitur First_genre
tf.fit(df['First_genre']) 
 
# Mapping array dari fitur index integer ke fitur Series_Title
tf.get_feature_names_out()

array(['action', 'adventure', 'animation', 'biography', 'comedy', 'crime',
       'drama', 'family', 'fantasy', 'film', 'horror', 'mystery', 'noir',
       'thriller', 'western'], dtype=object)

In [58]:
# Melakukan fit lalu ditransformasikan ke bentuk matrix
tfidf_matrix = tf.fit_transform(df["First_genre"]) 
 
# Melihat ukuran matrix tfidf
tfidf_matrix.shape

(1000, 15)

In [59]:
# Mengubah vektor tf-idf dalam bentuk matriks dengan fungsi todense()
tfidf_matrix.todense()

matrix([[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [1., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]])

In [60]:
# Membuat dataframe untuk melihat tf-idf matrix
# Kolom diisi dengan genre
# Baris diisi dengan judul film
 
pd.DataFrame(
    tfidf_matrix.todense(), 
    columns=tf.get_feature_names_out(),
    index=df.Series_Title
).sample(15, axis=1).sample(10, axis=0)

horror  comedy  film  mystery  thriller  western  \
Series_Title                                                                    
Bronenosets Potemkin            0.0     0.0   0.0      0.0       0.0      0.0   
Gandhi                          0.0     0.0   0.0      0.0       0.0      0.0   
Spoorloos                       0.0     0.0   0.0      1.0       0.0      0.0   
Dead Man's Shoes                0.0     0.0   0.0      0.0       0.0      0.0   
Uri: The Surgical Strike        0.0     0.0   0.0      0.0       0.0      0.0   
Into the Wild                   0.0     0.0   0.0      0.0       0.0      0.0   
The Best Years of Our Lives     0.0     0.0   0.0      0.0       0.0      0.0   
Goodfellas                      0.0     0.0   0.0      0.0       0.0      0.0   
Home Alone                      0.0     1.0   0.0      0.0       0.0      0.0   
Titanic                         0.0     0.0   0.0      0.0       0.0      0.0   

                             noir  action  adventure  family  animation  \
Series_Title                                                              
Bronenosets Potemkin          0.0     0.0        0.0     0.0        0.0   
Gandhi                        0.0     0.0        0.0     0.0        0.0   
Spoorloos                     0.0     0.0        0.0     0.0        0.0   
Dead Man's Shoes              0.0     0.0        0.0     0.0        0.0   
Uri: The Surgical Strike      0.0     1.0        0.0     0.0        0.0   
Into the Wild                 0.0     0.0        1.0     0.0        0.0   
The Best Years of Our Lives   0.0     0.0        0.0     0.0        0.0   
Goodfellas                    0.0     0.0        0.0     0.0        0.0   
Home Alone                    0.0     0.0        0.0     0.0        0.0   
Titanic                       0.0     0.0        0.0     0.0        0.0   

                             drama  crime  fantasy  biography  
Series_Title                                                   
Bronenosets Potemkin           1.0    0.0      0.0        0.0  
Gandhi                         0.0    0.0      0.0        1.0  
Spoorloos                      0.0    0.0      0.0        0.0  
Dead Man's Shoes               0.0    1.0      0.0        0.0  
Uri: The Surgical Strike       0.0    0.0      0.0        0.0  
Into the Wild                  0.0    0.0      0.0        0.0  
The Best Years of Our Lives    1.0    0.0      0.0        0.0  
Goodfellas                     0.0    0.0      0.0        1.0  
Home Alone                     0.0    0.0      0.0        0.0  
Titanic                        1.0    0.0      0.0        0.0

Gunakan cosine similarity untuk mengukur kemiripan antar film.

In [61]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
cos_sim = cosine_similarity(tfidf_matrix) 
cos_sim

array([[1., 0., 0., ..., 1., 1., 0.],
       [0., 1., 0., ..., 0., 0., 1.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [1., 0., 0., ..., 1., 1., 0.],
       [1., 0., 0., ..., 1., 1., 0.],
       [0., 1., 0., ..., 0., 0., 1.]])

In [62]:
# Membuat dataframe dari variabel cos_sim dengan baris dan kolom berupa judul film
cos_sim_dataframe = pd.DataFrame(cos_sim, index=df['Series_Title'], columns=df['Series_Title'])
print('Shape:', cos_sim_dataframe.shape)
 
# Melihat similarity matrix untuk tiap film
cos_sim_dataframe.sample(6, axis=1).sample(10, axis=0)

Shape: (1000, 1000)


Series_Title            Kind Hearts and Coronets  Brokeback Mountain  \
Series_Title                                                           
The Thin Man                                 1.0                 0.0   
The African Queen                            0.0                 0.0   
The Godfather                                0.0                 0.0   
Tumbbad                                      0.0                 1.0   
Nuovo Cinema Paradiso                        0.0                 1.0   
The Station Agent                            1.0                 0.0   
The Little Mermaid                           0.0                 0.0   
La piel que habito                           0.0                 1.0   
El secreto de sus ojos                       0.0                 1.0   
Hævnen                                       0.0                 1.0   

Series_Title            Planes, Trains & Automobiles  Sherlock Jr.  \
Series_Title                                                         
The Thin Man                                     1.0           0.0   
The African Queen                                0.0           0.0   
The Godfather                                    0.0           0.0   
Tumbbad                                          0.0           0.0   
Nuovo Cinema Paradiso                            0.0           0.0   
The Station Agent                                1.0           0.0   
The Little Mermaid                               0.0           0.0   
La piel que habito                               0.0           0.0   
El secreto de sus ojos                           0.0           0.0   
Hævnen                                           0.0           0.0   

Series_Title            Breaking the Waves  Bhaag Milkha Bhaag  
Series_Title                                                    
The Thin Man                           0.0                 0.0  
The African Queen                      0.0                 0.0  
The Godfather                          0.0                 0.0  
Tumbbad                                1.0                 0.0  
Nuovo Cinema Paradiso                  1.0                 0.0  
The Station Agent                      0.0                 0.0  
The Little Mermaid                     0.0                 0.0  
La piel que habito                     1.0                 0.0  
El secreto de sus ojos                 1.0                 0.0  
Hævnen                                 1.0                 0.0

In [63]:
def movie_recommendations(Series_Title, similarity_data=cos_sim_dataframe, features=df[["Series_Title", "First_genre"]], n=10):
    """
    Rekomendasi Film berdasarkan kemiripannya
 
    Parameter:
    ---
    judul_film : tipe data string (str)
                 judul film (index kemiripan dataframe)
    similarity_data : tipe data pd.DataFrame (object)
                      Kesamaan dataframe, simetrik, dengan film sebagai 
                      indeks dan kolom
    features : tipe data pd.DataFrame (object)
            Mengandung fitur-fitur yang digunakan untuk mendefinisikan kemiripan
    n : tipe data integer (int)
        Banyaknya jumlah rekomendasi yang diberikan
    """
 
 
    # Mengambil data dengan menggunakan argpartition untuk melakukan partisi secara tidak langsung sepanjang sumbu yang diberikan    
    # Dataframe diubah menjadi numpy
    # Range(start, stop, step)
    index = similarity_data.loc[:,Series_Title].to_numpy().argpartition(
        range(-1, -n, -1))
    
    # Mengambil data dengan similarity terbesar dari index yang ada
    nearest = similarity_data.columns[index[-1:-(n+2):-1]]
    
    # Drop nama_resto agar nama resto yang dicari tidak muncul dalam daftar rekomendasi
    nearest = nearest.drop(Series_Title, errors='ignore')
 
    return pd.DataFrame(nearest).merge(features).head(n)

# Evaluation

Mencari film yang mirip dan direkomendasikan setelah film X: First Class.

In [64]:
df[df.Series_Title.eq('X: First Class')]

Series_Title First_genre
749  X: First Class      Action

In [65]:
movie_recommendations('X: First Class')

Series_Title First_genre
0              Mission: Impossible - Fallout      Action
1                             V for Vendetta      Action
2  Tropa de Elite 2: O Inimigo Agora é Outro      Action
3                                    Dunkirk      Action
4                                   Deadpool      Action
5                                Udta Punjab      Action
6                                    Yip Man      Action
7                                    Warrior      Action
8                                      Raazi      Action
9                           Paan Singh Tomar      Action

Mencari film yang mirip dan direkomendasikan setelah film The Kid.

In [66]:
df[df.Series_Title.eq('The Kid')]

Series_Title First_genre
127      The Kid      Comedy

In [67]:
movie_recommendations('The Kid')

Series_Title First_genre
0                   Kind Hearts and Coronets      Comedy
1                                Barton Fink      Comedy
2                               Withnail & I      Comedy
3                   The Purple Rose of Cairo      Comedy
4                 The Muppet Christmas Carol      Comedy
5  Three Billboards Outside Ebbing, Missouri      Comedy
6                       Perfetti sconosciuti      Comedy
7                                After Hours      Comedy
8                                      Zelig      Comedy
9                                  Airplane!      Comedy